internal/incident/incidents.go

```diff
@@ -181,3 +238,99 @@ func ProcessEvent(
 
 	return currentIncident.ProcessEvent(ctx, ev, created)
 }
+
+// restoreAttrsFor restores the attributes of the given subject type from the database.
+//
+// It bulks SELECT the info from the database via a `SELECT ... WHERE ... IN(?)` query scoped to the specified args.
+// The column name for the where clause is determined automatically based on the provided subject type.
+// Currently, it only supports object.Object, object.IdTagRow, object.ExtraTagRow, EscalationState, ContactRow types
+// and panics if something else is provided.
+//
+// Note, this function accesses/modifies the specified incidents map without obtaining any locks
+// and must be protected against simultaneous write operations if a shared map is used.
+func restoreAttrsFor(ctx context.Context, db *icingadb.DB, incidents map[any]*Incident, subject any, scopes []any) error {
+	var (
+		// Name of the column used to filter the database query by.
+		column string
+		// A callback set based on the specified subject and called after each successful row retrieval.
+		restore func(any)
+		// A factory func set based on the given subject and called before each row scan.
+		factory func() any
+	)
+
+	switch subject.(type) {
```

Comment:
That can be done without the type switch. I've started a draft where one of the calls of this function is replaced with another function using generics instead: https://github.com/Icinga/icinga-notifications/commit/77f93ce291ee4e0ee6e15074cbb6f96536416416